### DISCUSSION TITLE
Why would file.close() freeze/crash pico using micropython?

### DISCUSSION THREAD

**Discussion by AngeloFatica** [micropython/micropython#14249] [RP2040 / Pico]
Hello,

I have had a freezing issue for a while now and I have found it to be caused by file.close() function when writing to a .txt file in my program. It seems to work sometimes but a good 33% of the time it will hang itself up on that portion specifically. The file is only around 25 lines and it is just storing 25 different numbers for me, so it is not large at all. I have tried using the 'with open' method with the same result. Is there something I need to be aware of when trying to close the file? Perhaps give it more time after the last item is written to close it? Right now it writes all items with a 100ms sleep in between each and a 100ms sleep between the last write and the close() call. This seems to only happen with the pico, not any stm I have tried. Any help is appreciated, thanks.

Edit: 
I am using a W5100S pico for this and have tried v1.19 to v1.22 micropython and the issue persists.

**Comment by mendenm** [comment]
Is it possible you have written files so many times, the built-in flash is dying?  I have a PicoW I was using as a data logger, which ran perfectly for 6 months, and then started freezing up frequently.  It would record data, and once a day it would be downloaded and erased. The write count on each block was probably a few hundred.  I'm pretty sure I have killed the flash.

**Reply by robert-hh** [reply]
> Is it possible you have written files so many times, the built-in flash is dying? 

It needs more than a few hundred erase/write cycles to damage a flash block. The vendors usually allow 100_000 cycles. And in normal operation conditions it needs more. In a test I made, it needed 5 Million writes. And VFS has wear level protection. FAT is a different story, because it overwrites the same blocks very often.

**Reply by AngeloFatica** [reply]
I do not believe so, as one of these sensor devices will only ever see a max of maybe 100 file read writes in their lifetime to be honest. I

**Reply by AngeloFatica** [reply]
I also have tried changing it to a JSON file instead of text and it seems to be behaving similarly. Perhaps it is an issue with memory fragmentation? Would the close function need a specific amount of unfragmented memory to perform? I have had issues in the past with fragmentation, but my free memory as it stands is above 100KB still. This is very strange behavior.

**Comment by AngeloFatica** [comment]
stack: 1796 out of 7936
GC: total: 193664, used: 107184, free: 86480
 No. of 1-blocks: 783, 2-blocks: 225, max blk sz: 1046, max free sz: 1002

This is my micropython.mem_info() print right before the freeze happens on file close. Is it an issue of something needing more than the max free block size? A fragmentation issue?

**Comment by AngeloFatica** [comment]
Update:

The issue is fixed by eliminating my threaded portion of the code altogether. I understand that _thread doesn't have the best track record with micropython, but I still am unsure as to why it would affect file system/flash at all if it is not using it or sharing any variables?

**Reply by Unchained-wp** [reply]
Hi, I have a similar issue. I start a loop on a new thread that samples a few things every 2 seconds and then save the data in json to a file every 5 minutes. The main thread runs a simple webserver to allow access to sampled values as needed. I've tried everything that I can think of to trace or eliminate the issue, like doing gc.collect() before and after the file saving. What is interesting is that if I print the progress during the saving process, it seems to run for longer before freezing but eventually also freezes, once again as it is closing the file.
The reason for saving the status regularly is to have a reasonably accurate restart position if failure should occur, but now the creation of that restart info causes the problem, a kind if catch 22.

I can try using an asynchronous webserver and use only one thread, but what is the use of having 2 cores and only one can be used in practice? I shall try to swap the core functionality around too so that the saving happens from main thread, but any other suggestions will be appreciated.

**Reply by AngeloFatica** [reply]
@Unchained-wp It seems like the threading hasn't been completely figured out yet with micropython as far as I've seen. I ended up changing my program to be single thread and have had absolutely zero issues anymore. Options as it stands seem to be use C if you want to use the other core or don't do file I/O with both cores enabled. Or you'll have to merge into a single thread.

**Reply by Unchained-wp** [reply]
@AngeloFatica Thank you for the info. I might just change to C/C++. I used Micropython because it is easy to use and to learn a bit of python but my primary experience is with C++. That should give me a slight speed advantage too.

**Reply by bixb922** [reply]
The file system that Micropython uses is Littlefs2 https://github.com/littlefs-project/littlefs

AFAIK Littlefs does not know about multithreading. As long as you do all file operation on the same thread or use mutex it should be safe.

**Reply by AngeloFatica** [reply]
@Unchained-wp same story for me. Micropython is super helpful to develop on quickly for proof of concept but does have its drawbacks like you mentioned the speed. But absolutely an awesome programming language nonetheless. 

@bixb922 I used the file I/O on the main core and it was still freezing on file closing calls. I never tried it with mutex though. Perhaps that could help. @Unchained-wp maybe you could give this a shot and see if it helps you out?

**Reply by Unchained-wp** [reply]
I only write to the file in one spot currently, don't even read it. I am actually using a mutex, the write follows immediately after I update the relevant variables if more than 5 minutes have passed since the previous write. The same mutex is also used when the webserver in the other thread reads the variables. I'll check the code again to see if I didn't perhaps missed using the mutex.

**Reply by MilhouseVH** [reply]
@AngeloFatica you may have experienced the issue discussed in https://github.com/micropython/micropython/issues/10448 - hopefully this is now fixed in 1.23.0, but if not then the solution (as you've discovered) is to avoid using both cores when using the WIZnet network module.

**Comment by peterhinch** [comment]
Re threading on RP2 you might want to read [this doc](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/THREADING.md). 

tl;dr It needs using with great care.
